File: cmd/subscribe.go
package cmd

import (
	"encoding/json"
	"fmt"
	"os"
	"time"

	"github.com/google/uuid"
	"github.com/mturley/agent-handler/config"
	"github.com/mturley/agent-handler/db"
	"github.com/mturley/agent-handler/worktreeinterop"
	"github.com/spf13/cobra"
)

var subscribeCmd = &cobra.Command{
	Use:   "subscribe",
	Short: "Subscribe a session to a resource",
	RunE:  runSubscribe,
}

var (
	subResource string
	subURL      string
)

func init() {
	subscribeCmd.GroupID = "agent"
	rootCmd.AddCommand(subscribeCmd)
	subscribeCmd.Flags().StringVar(&subResource, "resource", "", "resource ID (format: type:id, e.g., pr:owner/repo#42)")
	subscribeCmd.Flags().StringVar(&subURL, "url", "", "resource URL (optional)")
	subscribeCmd.Flags().String("session-id", "", "session ID (auto-detected if omitted)")
	subscribeCmd.Flags().Bool("related", false, "propagate to worktree as a related (not primary) resource")
	subscribeCmd.MarkFlagRequired("resource")
}

func runSubscribe(cmd *cobra.Command, args []string) error {
	d, err := openDB()
	if err != nil {
		return fmt.Errorf("failed to open database: %w", err)
	}
	defer d.Close()

	sessionID, err := resolveSessionID(cmd)
	if err != nil {
		return fmt.Errorf("could not determine session: %w", err)
	}

	// Parse resource
	resourceType, resourceID := worktreeinterop.ParseResourceID(subResource)
	if resourceType == "" {
		return fmt.Errorf("invalid resource format (expected type:id): %s", subResource)
	}

	// Check if the corresponding service is configured
	service := config.ResourceTypeToService(resourceType)
	if service != "" {
		if !config.ServiceConfiguredForWatching(service) {
			return fmt.Errorf("%s is not configured. Run 'handler watcher auth %s' to set up API access", service, service)
		}
	}

	// Auto-fill URL from config if not provided
	if subURL == "" {
		cfg, cfgErr := config.Read(config.DefaultPath())
		if cfgErr == nil && cfg != nil {
			subURL = cfg.DefaultResourceURL(resourceType, resourceID)
		}
	}

	// Subscribe
	now := time.Now().UTC().Format(time.RFC3339)
	var urlPtr *string
	if subURL != "" {
		urlPtr = &subURL
	}

	err = d.Subscribe(db.Subscription{
		ID:           uuid.New().String(),
		SessionID:    sessionID,
		ResourceType: resourceType,
		ResourceID:   resourceID,
		ResourceURL:  urlPtr,
		CreatedAt:    now,
	})
	if err != nil {
		return fmt.Errorf("failed to subscribe: %w", err)
	}

	// Propagate to the worktree CLI if present (best-effort).
	if worktreeinterop.Available() {
		related, _ := cmd.Flags().GetBool("related")
		cwd, _ := os.Getwd()
		u := ""
		if urlPtr != nil {
			u = *urlPtr
		}
		if err := worktreeinterop.AddResource(cwd, worktreeinterop.Resource{
			Type: resourceType, ID: resourceID, URL: u,
		}, related); err != nil {
			// soft degradation: subscribe already succeeded
			fmt.Fprintf(os.Stderr, "warning: could not propagate to worktree: %v\n", err)
		}
	}

	// Output
	if jsonOutput {
		output := map[string]interface{}{
			"session_id":    sessionID,
			"resource_type": resourceType,
			"resource_id":   resourceID,
			"resource_url":  subURL,
			"status":        "subscribed",
		}
		data, err := json.MarshalIndent(output, "", "  ")
		if err != nil {
			return err
		}
		fmt.Println(string(data))
	} else {
		fmt.Printf("✓ Subscribed session %s to %s:%s\n", sessionID, resourceType, resourceID)
		if subURL != "" {
			fmt.Printf("  URL: %s\n", subURL)
		}
	}

	return nil
}

File: src/Handler/Cmd/Subscribe.hs
-- | Port of cmd/subscribe.go, cmd/unsubscribe.go, cmd/subscriptions.go:
-- resource subscription management.
module Handler.Cmd.Subscribe
  ( subscribeCommand
  , unsubscribeCommand
  , subscriptionsCommand
  ) where

import Control.Monad (forM_, when)
import Data.Aeson (object, toJSON, (.=))
import Data.Maybe (isJust)
import Data.Text (Text)
import Options.Applicative

import Handler.Cli.Common
import Handler.Config
  ( configDefaultPath, defaultResourceUrl, isServiceConfigured
  , readConfig, resourceTypeToService )
import Handler.Db (close)
import Handler.Db.Subscriptions
  ( Subscription(..), listSubscriptions, subscribe, subscriptionToJson, unsubscribe )
import Handler.Util (newUuid, nowIso)
import Handler.Worktree (appendResource, parseResourceId, removeResource)

data SubscribeOpts = SubscribeOpts
  { sResource  :: Text
  , sUrl       :: Text
  , sSessionId :: Maybe Text
  , sPrimary   :: Bool
  , sPersist   :: Bool
  }

subscribeCommand :: Mod CommandFields NamedCommand
subscribeCommand = mkCommand "subscribe" "Subscribe a session to a resource" (runSubscribe <$> opts)
  where
    opts = SubscribeOpts
      <$> strOption (long "resource" <> metavar "RES"
            <> help "resource ID (format: type:id, e.g., pr:owner/repo#42)")
      <*> strOption (long "url" <> value "" <> help "resource URL (optional)")
      <*> sessionIdOption
      <*> switch (long "primary" <> help "mark as a primary resource for this worktree")
      <*> switch (long "persist" <> help "also write to .worktree-resources so future sessions in this worktree auto-subscribe")

runSubscribe :: SubscribeOpts -> Ctx -> IO ()
runSubscribe o ctx = do
  db <- openDb
  sessionId <- resolveSessionIdOpt o.sSessionId

  let (resourceType, resourceId) = parseResourceId o.sResource
  when (resourceType == "") $
    dieText ("invalid resource format (expected type:id): " <> o.sResource)

  -- Check if the corresponding service is configured
  cfgPath <- configDefaultPath
  cfg <- readConfig cfgPath
  let service = resourceTypeToService resourceType
  when (service /= "" && not (isServiceConfigured cfg service)) $
    dieText (service <> " is not configured. Run 'handler watcher auth " <> service <> "' to set up API access")

  -- Auto-fill URL from config if not provided
  let url = if o.sUrl == "" then defaultResourceUrl cfg resourceType resourceId else o.sUrl

  now <- nowIso
  subId <- newUuid
  subscribe db Subscription
    { subId = subId
    , sessionId = sessionId
    , resourceType = resourceType
    , resourceId = resourceId
    , resourceUrl = if url == "" then Nothing else Just url
    , createdAt = now
    , deletedAt = Nothing
    }

  when (o.sPersist && url /= "") $
    appendResource ".worktree-resources" o.sResource url o.sPrimary
  close db

  if ctx.jsonOutput
    then printJson $ object
      [ "session_id" .= sessionId
      , "resource_type" .= resourceType
      , "resource_id" .= resourceId
      , "resource_url" .= url
      , "status" .= ("subscribed" :: Text)
      ]
    else do
      printfT ["✓ Subscribed session ", sessionId, " to ", resourceType, ":", resourceId]
      when (url /= "") $ printfT ["  URL: ", url]

data UnsubscribeOpts = UnsubscribeOpts
  { unResource  :: Text
  , unSessionId :: Maybe Text
  , unPersist   :: Bool
  }

unsubscribeCommand :: Mod CommandFields NamedCommand
unsubscribeCommand = mkCommand "unsubscribe" "Unsubscribe a session from a resource" (runUnsubscribe <$> opts)
  where
    opts = UnsubscribeOpts
      <$> strOption (long "resource" <> metavar "RES" <> help "resource ID (format: type:id)")
      <*> sessionIdOption
      <*> switch (long "persist" <> help "also remove from .worktree-resources so future sessions won't auto-subscribe")

runUnsubscribe :: UnsubscribeOpts -> Ctx -> IO ()
runUnsubscribe o ctx = do
  db <- openDb
  sessionId <- resolveSessionIdOpt o.unSessionId

  let (resourceType, resourceId) = parseResourceId o.unResource
  when (resourceType == "") $
    dieText ("invalid resource format (expected type:id): " <> o.unResource)

  unsubscribe db sessionId resourceType resourceId

  when o.unPersist $
    removeResource ".worktree-resources" o.unResource
  close db

  if ctx.jsonOutput
    then printJson $ object
      [ "session_id" .= sessionId
      , "resource_type" .= resourceType
      , "resource_id" .= resourceId
      , "status" .= ("unsubscribed" :: Text)
      ]
    else printfT ["✓ Unsubscribed session ", sessionId, " from ", resourceType, ":", resourceId]

subscriptionsCommand :: Mod CommandFields NamedCommand
subscriptionsCommand = mkCommand "subscriptions" "List subscriptions for a session"
  (runSubscriptions <$> sessionIdOption
    <*> switch (long "all" <> help "include deleted subscriptions"))

runSubscriptions :: Maybe Text -> Bool -> Ctx -> IO ()
runSubscriptions msid includeDeleted ctx = do
  db <- openReadOnlyDb
  sessionId <- resolveSessionIdOpt msid
  subs <- listSubscriptions db sessionId includeDeleted
  close db

  if ctx.jsonOutput
    then printJson (toJSON (map subscriptionToJson subs))
    else if null subs
      then putTextLn "No subscriptions found"
      else do
        printfT ["Subscriptions for session ", sessionId, ":\n"]
        forM_ subs $ \sub -> do
          let status = if isJust sub.deletedAt then "deleted" else "active"
          printfT ["  ", sub.resourceType, ":", sub.resourceId, " [", status, "]"]
          forM_ sub.resourceUrl $ \u -> printfT ["    URL: ", u]
          printfT ["    Created: ", sub.createdAt]
          forM_ sub.deletedAt $ \d -> printfT ["    Deleted: ", d]
          putTextLn ""
